File: web_app/backend/engine/proofs.py
import math
import os

import numpy as np

from .rns import approx_center
from .sample import challenge, gaussian_int
from .util import chal_to_sparse, rej_accept, ser, sparse_mul
from .xof import XofReader, shake256

_RNG = np.random.default_rng()

REJ_ALPHA = 11.0
REJ_M = 3.0


def sigma_for(kappa, beta_inf, dim):
    return REJ_ALPHA * beta_inf * math.sqrt(kappa * dim)


def bits_of(sigma):
    return max(2, int(math.ceil(math.log2(6.0 * sigma))) + 1)


class Proof:
    def __init__(self, kind, data, size_bits):
        self.kind = kind
        self.data = data
        self.size_bits = size_bits


def _rnd(rnd):
    return rnd if rnd is not None else os.urandom(32)


def fwd_small(R, x):
    x = np.asarray(x, dtype=np.int64)
    xr = np.stack([x % p for p in R.primes], axis=-2)
    return R.fwd_many(xr, reduced=True)


def matvec_hat(R, Ah, xh):
    cols = Ah.shape[1]
    out = np.zeros((Ah.shape[0], R.L, R.n), dtype=np.int64)
    step = max(1, min(cols, R.lazy))
    for st in range(0, cols, step):
        acc = Ah[:, st] * xh[st]
        for j in range(st + 1, min(cols, st + step)):
            acc += Ah[:, j] * xh[j]
        out = (out + acc) % R.qs
    return out


def _dot_hat(R, bh, xh):
    cols = bh.shape[0]
    out = np.zeros((R.L, R.n), dtype=np.int64)
    step = max(1, min(cols, R.lazy))
    for st in range(0, cols, step):
        acc = bh[st] * xh[st]
        for j in range(st + 1, min(cols, st + step)):
            acc += bh[j] * xh[j]
        out = (out + acc) % R.qs
    return out


def prove_preimage(R, Ah, x, t, ctx, kappa, sigma, rnd=None, M=REJ_M, max_tries=60):
    k = x.shape[0]
    rnd = _rnd(rnd)
    for attempt in range(max_tries):
        y = gaussian_int(shake256(rnd, ctx, b"y", attempt), b"m", R.n, sigma, count=k)
        W = matvec_hat(R, Ah, fwd_small(R, y))
        dseed = shake256(ctx, ser(t), ser(W), b"chal")
        d = challenge(dseed, b"c", R.n, kappa)
        idx, sgn = chal_to_sparse(d)
        dx = sparse_mul(idx, sgn, x)
        z = y + dx
        if not rej_accept(z.reshape(-1), dx.reshape(-1), sigma, M, _RNG):
            continue
        if float(np.max(np.abs(z))) > 6.0 * sigma:
            continue
        size = k * R.n * bits_of(sigma) + 256
        return Proof("preimage", (dseed, z), size)
    return None


def verify_preimage(R, Ah, t, ctx, kappa, sigma, proof):
    if proof is None:
        return False
    dseed, z = proof.data
    if z.shape != (Ah.shape[1], R.n):
        return False
    if float(np.max(np.abs(z))) > 6.0 * sigma:
        return False
    d = challenge(dseed, b"c", R.n, kappa)
    dh = R.fwd(R.from_small(d))
    Az = matvec_hat(R, Ah, fwd_small(R, z))
    W = np.stack([(Az[i] - dh * t[i]) % R.qs for i in range(Az.shape[0])])
    return shake256(ctx, ser(t), ser(W), b"chal") == dseed


def prove_preimage_amortized(R, Ah, xs, ts, ctx, kappa, sigma, rnd=None, M=REJ_M,
                             max_tries=60):
    tau = len(xs)
    k = xs[0].shape[0]
    rnd = _rnd(rnd)
    for attempt in range(max_tries):
        y = gaussian_int(shake256(rnd, ctx, b"ay", attempt), b"m", R.n, sigma, count=k)
        W = matvec_hat(R, Ah, fwd_small(R, y))
        dseed = shake256(ctx, ser(*ts), ser(W), b"achal")
        acc = np.zeros((k, R.n), dtype=np.int64)
        for i in range(tau):
            d = challenge(shake256(dseed, b"i", i), b"c", R.n, kappa)
            idx, sgn = chal_to_sparse(d)
            acc += sparse_mul(idx, sgn, xs[i])
        z = y + acc
        if not rej_accept(z.reshape(-1), acc.reshape(-1), sigma, M, _RNG):
            continue
        if float(np.max(np.abs(z))) > 6.0 * sigma:
            continue
        size = k * R.n * bits_of(sigma) + 256
        return Proof("amortized", (dseed, z), size)
    return None


def verify_preimage_amortized(R, Ah, ts, ctx, kappa, sigma, proof):
    if proof is None:
        return False
    dseed, z = proof.data
    if float(np.max(np.abs(z))) > 6.0 * sigma:
        return False
    Az = matvec_hat(R, Ah, fwd_small(R, z))
    acc = np.zeros_like(Az)
    for i in range(len(ts)):
        d = challenge(shake256(dseed, b"i", i), b"c", R.n, kappa)
        dh = R.fwd(R.from_small(d))
        for row in range(Az.shape[0]):
            acc[row] = (acc[row] + dh * ts[i][row]) % R.qs
    W = np.stack([(Az[i] - acc[i]) % R.qs for i in range(Az.shape[0])])
    return shake256(ctx, ser(*ts), ser(W), b"achal") == dseed


def prove_lin(R, B1h, b2h, alpha_h, beta_h, c1, c2, c1p, c2p, r, rp, ctx,
              kappa, sigma, rnd=None, M=REJ_M, max_tries=60):
    k = B1h.shape[1]
    rnd = _rnd(rnd)
    for att in range(max_tries):
        yy = gaussian_int(shake256(rnd, ctx, b"ly", att), b"m", R.n, sigma, count=2 * k)
        y, yp = yy[:k], yy[k:]
        yyh = fwd_small(R, yy)
        yh, yph = yyh[:k], yyh[k:]
        t = matvec_hat(R, B1h, yh)
        tp = matvec_hat(R, B1h, yph)
        u = (alpha_h * _dot_hat(R, b2h, yh) - _dot_hat(R, b2h, yph)) % R.qs
        dseed = shake256(ctx, ser(c1, c2, c1p, c2p),
                         ser(np.stack([t[0], tp[0], u])), b"lin")
        d = challenge(dseed, b"c", R.n, kappa)
        idx, sgn = chal_to_sparse(d)
        dr = sparse_mul(idx, sgn, r)
        drp = sparse_mul(idx, sgn, rp)
        z, zp = y + dr, yp + drp
        if not rej_accept(np.concatenate([z.reshape(-1), zp.reshape(-1)]),
                          np.concatenate([dr.reshape(-1), drp.reshape(-1)]),
                          sigma, M, _RNG):
            continue
        if float(max(np.max(np.abs(z)), np.max(np.abs(zp)))) > 6.0 * sigma:
            continue
        return Proof("lin", (dseed, z, zp), 2 * k * R.n * bits_of(sigma) + 256)
    return None


def verify_lin(R, B1h, b2h, alpha_h, beta_h, c1, c2, c1p, c2p, ctx, kappa,
               sigma, proof):
    if proof is None:
        return False
    dseed, z, zp = proof.data
    if float(max(np.max(np.abs(z)), np.max(np.abs(zp)))) > 6.0 * sigma:
        return False
    d = challenge(dseed, b"c", R.n, kappa)
    dh = R.fwd(R.from_small(d))
    zzh = fwd_small(R, np.concatenate([z, zp]))
    zh, zph = zzh[:z.shape[0]], zzh[z.shape[0]:]
    t = (matvec_hat(R, B1h, zh) - dh * c1) % R.qs
    tp = (matvec_hat(R, B1h, zph) - dh * c1p) % R.qs
    rhs = (alpha_h * c2[0] + beta_h - c2p[0]) % R.qs
    u = (alpha_h * _dot_hat(R, b2h, zh) - _dot_hat(R, b2h, zph)
         - rhs * dh) % R.qs
    exp = shake256(ctx, ser(c1, c2, c1p, c2p),
                   ser(np.stack([t[0], tp[0], u])), b"lin")
    return exp == dseed


def gaussian_big_rns(R, seed, k, sigma):
    need = k * R.n
    rd = XofReader(seed, b"gbr")
    raw = np.frombuffer(rd.read(need * 16 + 8), dtype=np.uint64)[:2 * need]
    u1 = (raw[0::2] >> np.uint64(11)).astype(np.float64) / float(1 << 53)
    u2 = (raw[1::2] >> np.uint64(11)).astype(np.float64) / float(1 << 53)
    g = np.sqrt(-2.0 * np.log(np.clip(u1, 1e-300, None))) * np.cos(2.0 * np.pi * u2)
    t = max(0, int(math.floor(math.log2(max(sigma, 1.0)))) - 36)
    hi = np.rint(g * (sigma / float(1 << t))).astype(np.int64).reshape(k, R.n)
    lo = np.zeros((k, R.n), dtype=np.int64)
    if t:
        rawl = np.frombuffer(rd.read(need * 8 + 8), dtype=np.uint64)[:need]
        lo = (rawl % np.uint64(1 << t)).astype(np.int64).reshape(k, R.n)
    two_t = np.array([pow(2, t, p) for p in R.primes], dtype=np.int64)
    yr = np.zeros((k, R.L, R.n), dtype=np.int64)
    for i in range(R.L):
        yr[:, i, :] = (hi % R.primes[i] * two_t[i] + lo) % R.primes[i]
    return yr


def prove_bnd_amortized(R, Ah, xs_rns, ts, ctx, kappa, sigma, rnd=None, M=REJ_M,
                        max_tries=40):
    tau = len(xs_rns)
    k = xs_rns[0].shape[0]
    rnd = _rnd(rnd)
    for att in range(max_tries):
        yr = gaussian_big_rns(R, shake256(rnd, ctx, b"by", att), k, sigma)
        yh = R.fwd_many(yr)
        W = matvec_hat(R, Ah, yh)
        dseed = shake256(ctx, ser(*ts), ser(W), b"bchal")
        acc = np.zeros((k, R.L, R.n), dtype=np.int64)
        for i in range(tau):
            d = challenge(shake256(dseed, b"i", i), b"c", R.n, kappa)
            dh = R.fwd(R.from_small(d))
            acc = (acc + dh * R.fwd_many(xs_rns[i])) % R.qs
        accc = R.inv_many(acc)
        z = (yr + accc) % R.qs
        zf = np.concatenate([approx_center(R, z[j]) for j in range(k)])
        df = np.concatenate([approx_center(R, accc[j]) for j in range(k)])
        ex = math.exp(min(50.0, (-2.0 * float(np.dot(zf, df))
                                 + float(np.dot(df, df))) / (2.0 * sigma ** 2)))
        if _RNG.random() >= min(1.0, ex / M):
            continue
        if float(np.max(np.abs(zf))) > 6.0 * sigma:
            continue
        return Proof("bnd", (dseed, z), k * R.n * bits_of(sigma) + 256)
    return None


def verify_bnd_amortized(R, Ah, ts, ctx, kappa, sigma, proof):
    if proof is None:
        return False
    dseed, z = proof.data
    zf = np.concatenate([approx_center(R, z[j]) for j in range(z.shape[0])])
    if float(np.max(np.abs(zf))) > 6.0 * sigma:
        return False
    zh = R.fwd_many(z)
    Az = matvec_hat(R, Ah, zh)
    acc = np.zeros_like(Az)
    for i in range(len(ts)):
        d = challenge(shake256(dseed, b"i", i), b"c", R.n, kappa)
        dh = R.fwd(R.from_small(d))
        for row in range(Az.shape[0]):
            acc[row] = (acc[row] + dh * ts[i][row]) % R.qs
    W = np.stack([(Az[i] - acc[i]) % R.qs for i in range(Az.shape[0])])
    return shake256(ctx, ser(*ts), ser(W), b"bchal") == dseed

File: android_app/app/src/main/java/org/oylama/crypto/Proofs.java
package org.oylama.crypto;

import java.util.Arrays;
import java.util.concurrent.ThreadLocalRandom;

public final class Proofs {
    public static final double REJ_ALPHA = 11.0;
    public static final double REJ_M = 3.0;

    private Proofs() {
    }

    public static double sigmaFor(int kappa, double betaInf, long dim) {
        return REJ_ALPHA * betaInf * Math.sqrt((double) kappa * dim);
    }

    public static int bitsOf(double sigma) {
        return Math.max(2, (int) Math.ceil(Math.log(6.0 * sigma) / Math.log(2.0)) + 1);
    }

    static long[][] add(long[][] a, long[][] b) {
        long[][] r = new long[a.length][];
        for (int i = 0; i < a.length; i++) {
            r[i] = new long[a[i].length];
            for (int k = 0; k < a[i].length; k++) {
                r[i][k] = a[i][k] + b[i][k];
            }
        }
        return r;
    }

    static double maxAbs(long[][] z) {
        long m = 0;
        for (long[] row : z) {
            for (long v : row) {
                long a = Math.abs(v);
                if (a > m) {
                    m = a;
                }
            }
        }
        return m;
    }

    static boolean rejAccept(long[][] z, long[][] cx, double sigma, double M) {
        double dot = 0;
        double nrm = 0;
        for (int i = 0; i < z.length; i++) {
            for (int k = 0; k < z[i].length; k++) {
                double c = cx[i][k];
                dot += (double) z[i][k] * c;
                nrm += c * c;
            }
        }
        double ex = Math.exp(Math.min(50.0, (-2.0 * dot + nrm) / (2.0 * sigma * sigma)));
        return ThreadLocalRandom.current().nextDouble() < Math.min(1.0, ex / M);
    }

    static int[] subMul(Ring R, int[] a, int[] dh, int[] t) {
        return R.sub(a, R.mulHat(dh, t));
    }

    public static Proof provePreimage(Ring R, int[][][] Ah, long[][] x, int[][] t, byte[] ctx, int kappa, double sigma, byte[] rnd) {
        int k = x.length;
        Hash.Stream pre = Hash.stream(ctx, Hash.ser(t));
        for (int att = 0; att < 60; att++) {
            long[][] y = Sample.gaussian(Hash.h(rnd, ctx, "y", att), "m", R.n, sigma, k);
            int[][] W = R.matvecHat(Ah, R.fwdSmall(y));
            byte[] dseed = pre.copy().add(Hash.ser(W)).add("chal").digest(32);
            Sample.Sparse d = Sample.challenge(dseed, "c", R.n, kappa);
            long[][] dx = Sample.sparseMul(d, x);
            long[][] z = add(y, dx);
            if (!rejAccept(z, dx, sigma, REJ_M)) {
                continue;
            }
            if (maxAbs(z) > 6.0 * sigma) {
                continue;
            }
            return Proof.preimage(dseed, z, (long) k * R.n * bitsOf(sigma) + 256);
        }
        return null;
    }

    public static boolean verifyPreimage(Ring R, int[][][] Ah, int[][] t, byte[] ctx, int kappa, double sigma, Proof p) {
        if (p == null || p.z == null || p.z.length != Ah[0].length) {
            return false;
        }
        for (long[] row : p.z) {
            if (row.length != R.n) {
                return false;
            }
        }
        if (maxAbs(p.z) > 6.0 * sigma) {
            return false;
        }
        Sample.Sparse d = Sample.challenge(p.seed, "c", R.n, kappa);
        int[] dh = R.fwd(R.fromSmall(Sample.dense(d, R.n)));
        int[][] Az = R.matvecHat(Ah, R.fwdSmall(p.z));
        int[][] W = new int[Az.length][];
        for (int r = 0; r < Az.length; r++) {
            W[r] = subMul(R, Az[r], dh, t[r]);
        }
        return Arrays.equals(Hash.h(ctx, Hash.ser(t), Hash.ser(W), "chal"), p.seed);
    }

    static Hash.Ser serAll(int[][][] ts) {
        int total = 0;
        for (int[][] t : ts) {
            total += t.length;
        }
        int[][] all = new int[total][];
        int o = 0;
        for (int[][] t : ts) {
            for (int[] row : t) {
                all[o++] = row;
            }
        }
        return Hash.ser(all);
    }

    public static Proof provePreimageAmortized(Ring R, int[][][] Ah, long[][][] xs, int[][][] ts, byte[] ctx, int kappa, double sigma, byte[] rnd) {
        int k = xs[0].length;
        Hash.Stream pre = Hash.stream(ctx, serAll(ts));
        for (int att = 0; att < 60; att++) {
            long[][] y = Sample.gaussian(Hash.h(rnd, ctx, "ay", att), "m", R.n, sigma, k);
            int[][] W = R.matvecHat(Ah, R.fwdSmall(y));
            byte[] dseed = pre.copy().add(Hash.ser(W)).add("achal").digest(32);
            long[][] acc = new long[k][R.n];
            for (int i = 0; i < xs.length; i++) {
                Sample.Sparse d = Sample.challenge(Hash.h(dseed, "i", i), "c", R.n, kappa);
                long[][] m = Sample.sparseMul(d, xs[i]);
                for (int j = 0; j < k; j++) {
                    for (int q = 0; q < R.n; q++) {
                        acc[j][q] += m[j][q];
                    }
                }
            }
            long[][] z = add(y, acc);
            if (!rejAccept(z, acc, sigma, REJ_M)) {
                continue;
            }
            if (maxAbs(z) > 6.0 * sigma) {
                continue;
            }
            return new Proof("amortized", dseed, z, null, null, (long) k * R.n * bitsOf(sigma) + 256);
        }
        return null;
    }

    public static boolean verifyPreimageAmortized(Ring R, int[][][] Ah, int[][][] ts, byte[] ctx, int kappa, double sigma, Proof p) {
        if (p == null) {
            return false;
        }
        if (maxAbs(p.z) > 6.0 * sigma) {
            return false;
        }
        int[][] Az = R.matvecHat(Ah, R.fwdSmall(p.z));
        int rows = Az.length;
        int[][] acc = new int[rows][R.L * R.n];
        for (int i = 0; i < ts.length; i++) {
            Sample.Sparse d = Sample.challenge(Hash.h(p.seed, "i", i), "c", R.n, kappa);
            int[] dh = R.fwd(R.fromSmall(Sample.dense(d, R.n)));
            for (int r = 0; r < rows; r++) {
                R.addMulHat(acc[r], dh, ts[i][r]);
            }
        }
        int[][] W = new int[rows][];
        for (int r = 0; r < rows; r++) {
            W[r] = R.sub(Az[r], acc[r]);
        }
        return Arrays.equals(Hash.h(ctx, serAll(ts), Hash.ser(W), "achal"), p.seed);
    }

    static Hash.Ser serLin(int[][] c1, int[][] c2, int[][] c1p, int[][] c2p) {
        int[][] all = new int[c1.length + c2.length + c1p.length + c2p.length][];
        int o = 0;
        for (int[][] g : new int[][][]{c1, c2, c1p, c2p}) {
            for (int[] row : g) {
                all[o++] = row;
            }
        }
        return Hash.ser(all);
    }

    static long[][] concat(long[][] a, long[][] b) {
        long[][] r = new long[a.length + b.length][];
        System.arraycopy(a, 0, r, 0, a.length);
        System.arraycopy(b, 0, r, a.length, b.length);
        return r;
    }

    public static Proof proveLin(Ring R, int[][][] B1h, int[][] b2h, int[] alpha, int[] beta, int[][] c1, int[][] c2,
                                 int[][] c1p, int[][] c2p, long[][] r, long[][] rp, byte[] ctx, int kappa, double sigma, byte[] rnd) {
        int k = B1h[0].length;
        Hash.Stream pre = Hash.stream(ctx, serLin(c1, c2, c1p, c2p));
        for (int att = 0; att < 60; att++) {
            long[][] yy = Sample.gaussian(Hash.h(rnd, ctx, "ly", att), "m", R.n, sigma, 2 * k);
            long[][] y = Arrays.copyOfRange(yy, 0, k);
            long[][] yp = Arrays.copyOfRange(yy, k, 2 * k);
            int[][] yh = R.fwdSmall(y);
            int[][] yph = R.fwdSmall(yp);
            int[][] t = R.matvecHat(B1h, yh);
            int[][] tp = R.matvecHat(B1h, yph);
            int[] u = R.sub(R.mulHat(alpha, R.dotHat(b2h, yh)), R.dotHat(b2h, yph));
            byte[] dseed = pre.copy().add(Hash.ser(t[0], tp[0], u)).add("lin").digest(32);
            Sample.Sparse d = Sample.challenge(dseed, "c", R.n, kappa);
            long[][] dr = Sample.sparseMul(d, r);
            long[][] drp = Sample.sparseMul(d, rp);
            long[][] z = add(y, dr);
            long[][] zp = add(yp, drp);
            if (!rejAccept(concat(z, zp), concat(dr, drp), sigma, REJ_M)) {
                continue;
            }
            if (Math.max(maxAbs(z), maxAbs(zp)) > 6.0 * sigma) {
                continue;
            }
            return new Proof("lin", dseed, z, zp, null, 2L * k * R.n * bitsOf(sigma) + 256);
        }
        return null;
    }

    public static boolean verifyLin(Ring R, int[][][] B1h, int[][] b2h, int[] alpha, int[] beta, int[][] c1, int[][] c2,
                                    int[][] c1p, int[][] c2p, byte[] ctx, int kappa, double sigma, Proof p) {
        if (p == null) {
            return false;
        }
        if (Math.max(maxAbs(p.z), maxAbs(p.zp)) > 6.0 * sigma) {
            return false;
        }
        Sample.Sparse d = Sample.challenge(p.seed, "c", R.n, kappa);
        int[] dh = R.fwd(R.fromSmall(Sample.dense(d, R.n)));
        int[][] zh = R.fwdSmall(p.z);
        int[][] zph = R.fwdSmall(p.zp);
        int[] t = R.sub(R.matvecHat(B1h, zh)[0], R.mulHat(dh, c1[0]));
        int[] tp = R.sub(R.matvecHat(B1h, zph)[0], R.mulHat(dh, c1p[0]));
        int[] rhs = R.sub(R.add(R.mulHat(alpha, c2[0]), beta), c2p[0]);
        int[] u = R.sub(R.sub(R.mulHat(alpha, R.dotHat(b2h, zh)), R.dotHat(b2h, zph)), R.mulHat(rhs, dh));
        return Arrays.equals(Hash.h(ctx, serLin(c1, c2, c1p, c2p), Hash.ser(t, tp, u), "lin"), p.seed);
    }

    public static int[][] gaussianBigRns(Ring R, byte[] seed, int k, double sigma) {
        int n = R.n;
        int need = k * n;
        Xof rd = new Xof(seed, "gbr");
        byte[] raw = rd.read(need * 16 + 8);
        int t = Math.max(0, (int) Math.floor(Math.log(Math.max(sigma, 1.0)) / Math.log(2.0)) - 36);
        double scale = sigma / (double) (1L << t);
        long[] hi = new long[need];
        for (int i = 0; i < need; i++) {
            long a = le64(raw, 16 * i);
            long b = le64(raw, 16 * i + 8);
            double u1 = (a >>> 11) / 9007199254740992.0;
            double u2 = (b >>> 11) / 9007199254740992.0;
            double g = Math.sqrt(-2.0 * Math.log(Math.max(u1, 1e-300))) * Math.cos(2.0 * Math.PI * u2);
            hi[i] = (long) Math.rint(g * scale);
        }
        long[] lo = new long[need];
        if (t > 0) {
            byte[] rawl = rd.read(need * 8 + 8);
            long mask = (1L << t) - 1;
            for (int i = 0; i < need; i++) {
                lo[i] = le64(rawl, 8 * i) & mask;
            }
        }
        int[][] yr = new int[k][R.L * n];
        for (int i = 0; i < R.L; i++) {
            long p = R.q[i];
            long twoT = Ntt.powmod(2, t, p);
            for (int j = 0; j < need; j++) {
                long v = (Math.floorMod(hi[j], p) * twoT + lo[j]) % p;
                yr[j / n][i * n + (j % n)] = (int) v;
            }
        }
        return yr;
    }

    static long le64(byte[] a, int o) {
        return (a[o] & 0xFFL) | ((a[o + 1] & 0xFFL) << 8) | ((a[o + 2] & 0xFFL) << 16) | ((a[o + 3] & 0xFFL) << 24)
            | ((a[o + 4] & 0xFFL) << 32) | ((a[o + 5] & 0xFFL) << 40) | ((a[o + 6] & 0xFFL) << 48) | ((a[o + 7] & 0xFFL) << 56);
    }

    static double[] approxAll(Ring R, int[][] z) {
        double[] out = new double[z.length * R.n];
        for (int j = 0; j < z.length; j++) {
            System.arraycopy(R.approxCenter(z[j]), 0, out, j * R.n, R.n);
        }
        return out;
    }

    public static Proof proveBndAmortized(Ring R, int[][][] Ah, int[][][] xs, int[][][] ts, byte[] ctx, int kappa, double sigma, byte[] rnd) {
        int k = xs[0].length;
        Hash.Stream pre = Hash.stream(ctx, serAll(ts));
        for (int att = 0; att < 40; att++) {
            int[][] yr = gaussianBigRns(R, Hash.h(rnd, ctx, "by", att), k, sigma);
            int[][] W = R.matvecHat(Ah, R.fwdAll(yr));
            byte[] dseed = pre.copy().add(Hash.ser(W)).add("bchal").digest(32);
            int[][] acc = new int[k][R.L * R.n];
            for (int i = 0; i < xs.length; i++) {
                Sample.Sparse d = Sample.challenge(Hash.h(dseed, "i", i), "c", R.n, kappa);
                int[] dh = R.fwd(R.fromSmall(Sample.dense(d, R.n)));
                for (int j = 0; j < k; j++) {
                    R.addMulHat(acc[j], dh, R.fwd(xs[i][j]));
                }
            }
            int[][] accc = R.invAll(acc);
            int[][] z = new int[k][];
            for (int j = 0; j < k; j++) {
                z[j] = R.add(yr[j], accc[j]);
            }
            double[] zf = approxAll(R, z);
            double[] df = approxAll(R, accc);
            double dot = 0, nrm = 0, zmax = 0;
            for (int i = 0; i < zf.length; i++) {
                dot += zf[i] * df[i];
                nrm += df[i] * df[i];
                zmax = Math.max(zmax, Math.abs(zf[i]));
            }
            double ex = Math.exp(Math.min(50.0, (-2.0 * dot + nrm) / (2.0 * sigma * sigma)));
            if (ThreadLocalRandom.current().nextDouble() >= Math.min(1.0, ex / REJ_M)) {
                continue;
            }
            if (zmax > 6.0 * sigma) {
                continue;
            }
            return new Proof("bnd", dseed, null, null, z, (long) k * R.n * bitsOf(sigma) + 256);
        }
        return null;
    }

    public static boolean verifyBndAmortized(Ring R, int[][][] Ah, int[][][] ts, byte[] ctx, int kappa, double sigma, Proof p) {
        if (p == null || p.zr == null) {
            return false;
        }
        double[] zf = approxAll(R, p.zr);
        for (double v : zf) {
            if (Math.abs(v) > 6.0 * sigma) {
                return false;
            }
        }
        int[][] Az = R.matvecHat(Ah, R.fwdAll(p.zr));
        int rows = Az.length;
        int[][] acc = new int[rows][R.L * R.n];
        for (int i = 0; i < ts.length; i++) {
            Sample.Sparse d = Sample.challenge(Hash.h(p.seed, "i", i), "c", R.n, kappa);
            int[] dh = R.fwd(R.fromSmall(Sample.dense(d, R.n)));
            for (int r = 0; r < rows; r++) {
                R.addMulHat(acc[r], dh, ts[i][r]);
            }
        }
        int[][] W = new int[rows][];
        for (int r = 0; r < rows; r++) {
            W[r] = R.sub(Az[r], acc[r]);
        }
        return Arrays.equals(Hash.h(ctx, serAll(ts), Hash.ser(W), "bchal"), p.seed);
    }
}
